File: go/Largest_Rectangle_in_Histogram.go
package main

func largestRectangleArea(ht []int) int {
	ht = append(ht, 0)
	ans := 0
	var stk []int
	for i, h := range ht {
		for len(stk) > 0 && h < ht[stk[len(stk)-1]] {
			top := stk[len(stk)-1]
			stk = stk[:len(stk)-1]
			l := i
			if len(stk) > 0 {
				l = i - stk[len(stk)-1] - 1
			}
			ans = max(ans, ht[top]*l)
		}
		stk = append(stk, i)
	}
	return ans
}

File: cpp/Largest_Rectangle_in_Histogram.cpp
#include <algorithm>
#include <stack>
#include <vector>
using namespace std;

class Solution {
public:
  int largestRectangleArea(vector<int> &h) {
    h.push_back(0);
    int ans = 0;
    stack<int> s;
    for (int i = 0; i < h.size(); i++) {
      while (!s.empty() && h[s.top()] > h[i]) {
        int top = s.top();
        s.pop();
        ans = max(ans, h[top] * (s.empty() ? i : (i - s.top() - 1)));
      }
      s.push(i);
    }
    return ans;
  }
};
